bug: timeout handler overrides capacity error detection causing false workflow failures

## Problem Summary

The OCI workflow is incorrectly reporting failures when Oracle Cloud capacity/limit errors occur, contradicting the documented expected behavior in CLAUDE.md. A race condition exists between the timeout handler and capacity error detection logic.

## Root Cause Analysis

### Technical Issue
A critical bug in `scripts/launch-parallel.sh` where the 55-second timeout handler overrides legitimate capacity error exit codes, causing false workflow failures.

**Sequence of Events:**
1. Both VM.Standard.A1.Flex and VM.Standard.E2.1.Micro instances encounter capacity issues
2. E2.1.Micro fails immediately with `LimitExceeded` (should exit code 2)  
3. A1.Flex fails with `InternalError: Out of host capacity` (should exit code 2)
4. E2.1.Micro verification process continues running (5 checks × 30s intervals)
5. 55-second timeout is reached while verification is still active
6. Timeout handler overwrites both exit codes to `EXIT_TIMEOUT_ERROR` instead of preserving `OCI_EXIT_CAPACITY_ERROR`
7. Result evaluation logic incorrectly categorizes this as "configuration or authentication errors"

### Code Location
**File:** `scripts/launch-parallel.sh`  
**Lines:** ~227-230, 275-343

```bash
# CURRENT PROBLEMATIC CODE:
if [[ $elapsed -ge $timeout_seconds ]]; then
    log_warning "Execution timeout reached (${timeout_seconds}s) - terminating background processes"
    terminate_processes
    STATUS_A1=$EXIT_TIMEOUT_ERROR  # BUG: Overwrites capacity error codes
    STATUS_E2=$EXIT_TIMEOUT_ERROR  # BUG: Overwrites capacity error codes
fi
```

## Evidence - Workflow Run #17358848061

**Date:** 2025-08-31T15:08:42Z  
**Duration:** 1m 34s  
**Result:** ❌ Failed (should be ✅ Success)

### Error Details:
```json
# E2.1.Micro Error:
{
    "code": "LimitExceeded", 
    "message": "The following service limits were exceeded: standard-e2-micro-core-count",
    "status": 400
}

# A1.Flex Error:
{
    "code": "InternalError",
    "message": "Out of host capacity", 
    "status": 500
}
```

### Log Evidence:
```
[INFO] LimitExceeded error in AD fgaj:AP-SINGAPORE-1-AD-1 (attempt 1/1)
[INFO] No capacity available for shape in AD fgaj:AP-SINGAPORE-1-AD-1 (attempt 1/1)
[WARNING] Execution timeout reached (55s) - terminating background processes
[ERROR] Parallel execution failed: Instance creation failed due to configuration or authentication errors
```

## Expected vs Actual Behavior

### Expected (per CLAUDE.md):
> **Capacity errors are EXPECTED** (treat as success - retry on schedule)  
> **"Out of host capacity"** - Common during peak usage periods

**Should result in:**
- Workflow status: ✅ **Success**
- Exit code: `0`
- Log message: "Both shapes unavailable due to capacity/limits - will retry on next schedule"

### Actual:
- Workflow status: ❌ **Failed**
- Exit code: `1` 
- Log message: "Instance creation failed due to configuration or authentication errors"

## Impact Assessment

### Critical Issues:
1. **False Failures:** 261 workflow runs show misleading failure status
2. **Notification Pollution:** Potential false failure alerts sent to users
3. **Success Tracking Corruption:** `SUCCESS_TRACKING_ENABLED` data becomes unreliable
4. **Monitoring Confusion:** Capacity exhaustion appears as system errors
5. **Resource Waste:** Manual investigation of "failures" that are actually expected behavior

### Pattern Analysis:
Recent runs show consistent pattern of capacity-related "failures" during peak Oracle Cloud usage periods, exactly matching documented expected behavior.

## Proposed Solution

### 1. Preserve Original Exit Codes
```bash
# Store exit codes before timeout handling
ORIGINAL_STATUS_A1=$STATUS_A1
ORIGINAL_STATUS_E2=$STATUS_E2

if [[ $elapsed -ge $timeout_seconds ]]; then
    log_warning "Execution timeout reached (${timeout_seconds}s) - terminating background processes"
    terminate_processes
    
    # Only override if processes hadn't already failed with specific error codes
    [[ $ORIGINAL_STATUS_A1 -eq -1 ]] && STATUS_A1=$EXIT_TIMEOUT_ERROR || STATUS_A1=$ORIGINAL_STATUS_A1
    [[ $ORIGINAL_STATUS_E2 -eq -1 ]] && STATUS_E2=$EXIT_TIMEOUT_ERROR || STATUS_E2=$ORIGINAL_STATUS_E2
fi
```

### 2. Enhanced Logging
Add timeout context to distinguish between:
- Pure timeout (no result files)
- Capacity error + timeout (result files exist with exit code 2)
- Mixed scenarios

### 3. Result File Timeout Handling  
Implement more robust result file reading with proper timeout and error code preservation.

## Files Affected
- `scripts/launch-parallel.sh` (primary)
- Potentially `scripts/utils.sh` (exit code constants)

## Testing Strategy

### Test Cases:
1. **Pure Capacity Errors:** Both shapes fail quickly with capacity issues
2. **Timeout + Capacity:** One shape has capacity error, other times out during verification
3. **Pure Timeout:** Both shapes timeout without result files
4. **Mixed Scenarios:** Various combinations of error types

### Validation:
- Verify exit code 0 when both shapes fail due to capacity (`LimitExceeded`, `InternalError`)
- Confirm proper logging shows capacity issues, not config errors
- Test that timeout handling preserves existing error classifications
- Ensure success tracking works correctly

## Priority Justification

**High Priority** due to:
- Affects core workflow reliability measurement
- Creates false operational alerts  
- Contradicts documented system behavior
- Impacts automated scheduling decisions
- Wastes engineering time on false failure investigations

## Related Documentation
- **CLAUDE.md:** Lines 38-42 - "Capacity errors are EXPECTED"
- **CLAUDE.md:** Lines 44-46 - Oracle Cloud capacity patterns
- **Infrastructure-deployment.yml:** Lines 214-237 - Failure notification logic

--- a/scripts/launch-parallel.sh
+++ b/scripts/launch-parallel.sh
@@ -370,14 +370,6 @@ main() {
         ((elapsed += sleep_interval))
     done
 
-    # Handle timeout case
-    if [[ $elapsed -ge $timeout_seconds ]]; then
-        log_warning "Execution timeout reached (${timeout_seconds}s) - terminating background processes"
-        terminate_processes
-        STATUS_A1=$EXIT_TIMEOUT_ERROR # Standard timeout exit code (GNU timeout compatibility)
-        STATUS_E2=$EXIT_TIMEOUT_ERROR # Standard timeout exit code (GNU timeout compatibility)
-    fi
-
     # Always wait for processes to fully complete and flush their output (handle empty PIDs)
     if [[ -n "$PID_A1" ]]; then
         wait $PID_A1 2>/dev/null || true
@@ -399,6 +391,37 @@ main() {
         log_debug "E2 result file found with status: $STATUS_E2"
     else
         log_warning "E2 result file not found - using default failure status"
+    fi
+    # Handle timeout case - architecture-aware approach respecting smart shape filtering
+    if [[ $elapsed -ge $timeout_seconds ]]; then
+        log_warning "Execution timeout reached (${timeout_seconds}s) - terminating background processes"
+        terminate_processes
+        
+        # Only apply timeout errors to shapes that were actually launched and have generic error codes
+        # This preserves capacity/limit error codes (2, 5) which indicate expected Oracle Cloud behavior
+        if [[ "$should_launch_a1" == true ]]; then
+            # Only override if no specific error code was already captured
+            if [[ $STATUS_A1 -eq 1 ]]; then
+                STATUS_A1=$EXIT_TIMEOUT_ERROR
+                log_debug "A1 timeout applied (was launched, no specific error code)"
+            else
+                log_debug "A1 timeout occurred but preserving error code $STATUS_A1 (capacity/limit detection)"
+            fi
+        else
+            log_debug "A1 was skipped due to cached limits - no timeout handling needed"
+        fi
+        
+        if [[ "$should_launch_e2" == true ]]; then
+            # Only override if no specific error code was already captured
+            if [[ $STATUS_E2 -eq 1 ]]; then
+                STATUS_E2=$EXIT_TIMEOUT_ERROR
+                log_debug "E2 timeout applied (was launched, no specific error code)"
+            else
+                log_debug "E2 timeout occurred but preserving error code $STATUS_E2 (capacity/limit detection)"
+            fi
+        else
+            log_debug "E2 was skipped due to cached limits - no timeout handling needed"
+        fi
     fi
     
     # Verify and update state for both instances (if state management enabled)
@@ -423,7 +446,6 @@ main() {
             log_debug "Skipping verification - instances were served from cache"
         fi
     fi
-    
     # Cleanup temporary files
     rm -rf "$temp_dir" 2>/dev/null || true
 

